### DISCUSSION TITLE
Performances issues using the DevTools

### DISCUSSION THREAD

**Discussion by astahmer** [TanStack/query#3932] [Q&A]
hey, first I wanna say thank you for your awesome work on everything related to the tanstack, it has been a wonder to work with daily

i'm currently updating lots of dependencies in a private work repo, and i'm facing performances issues that were not there before ("react-query": "4.0.0-alpha.21", or even before that with "react-query": "^3.34.8",)

when using the dev tools, it seems like there is a huge amount of work (re-renders from QueryRow ?) done by the DevTools despite it being closed 

looking at the [source](https://github.com/TanStack/query/blob/main/packages/react-query-devtools/src/devtools.tsx#L247), I think a naive solution would be to add a render condition like `{isResolvedOpen ? <ReactQueryDevtoolsPanel ...> (...)</ReactQueryDevtoolsPanel> : null`

without the dev tools:
![Screenshot 2022-07-28 at 15 13 05](https://user-images.githubusercontent.com/47224540/181514269-721bf0c3-1642-4bee-bd1f-89d026bb16ec.png)

with the dev tools:
![Screenshot 2022-07-28 at 15 14 23](https://user-images.githubusercontent.com/47224540/181514324-4e44ba81-03a0-4ddc-bc27-73414f3e07c9.png)
![Screenshot 2022-07-28 at 15 21 37](https://user-images.githubusercontent.com/47224540/181515682-2c36fe52-4c5d-4e34-9b73-6dedcdd1ca3d.png)
![Screenshot 2022-07-28 at 15 21 47](https://user-images.githubusercontent.com/47224540/181515761-26135895-a8d3-424c-91f8-a4d28ecf0526.png)

do you have any idea why/where that could be coming from ?

on a side note, I also get quite a lot of these since the update:
![Screenshot 2022-07-28 at 15 16 30](https://user-images.githubusercontent.com/47224540/181514451-e1ad53e4-bd43-4b41-a6ea-2f6b11dd3c1c.png)

**Comment by TkDodo** [comment]
we've added lots of improvements in this regard with v4, starting with the `beta`. Please use at least `react-query: "4.0.0-beta.23"`, or at best, update to the stable v4, which you can find in `@tanstack/react-query: "4.0.10"`

**Reply by astahmer** [reply]
I'm sorry I completely forgot to specify that I'm currently using `"@tanstack/react-query": "^4.0.5",`, updating from `("react-query": "4.0.0-alpha.21")` where I had no issues

I'll try updating right now to `@tanstack/react-query: "4.0.10"` and see how it goes !

**Reply by TkDodo** [reply]
hm, there won't be much difference between these two versions I'm afraid. can you try to reproduce this in codesandbox?

we actually fixed an issue in v4 that should've made this better:
- https://github.com/TanStack/query/issues/2742

but we also removed the workaround, so maybe not 🤔

**Reply by astahmer** [reply]
I did try to reproduce it using the codesandbox template but unfortunately I had no perfs issues there.. 
I suppose it makes sense since that template is quite a simple app. It might also be worth mentioning that I haven't yet updated react to v18, could that be related ?

**Reply by TkDodo** [reply]
> I haven't yet updated react to v18, could that be related ?

i hope not. or maybe, because we use the `useSyncExternalStore` shim under the hood.

> I suppose it makes sense since that template is quite a simple app

yeah that's true - what you would need to do is add code that goes into the direction of your app, so more queries or whatever it is that you are doing, until we have a minimal repro that shows the issue. Maybe you have lots of queries that mount at the same time etc ...
